system/jlib/jfile.cpp

```diff
@@ -7579,3 +7589,103 @@ IAPICopyClient * createApiCopyClient(IStorageApiInfo * source, IStorageApiInfo *
     }
     return nullptr;
 }
+
+
+class CBlockedFileIO : public CSimpleInterfaceOf<IFileIO>
+{
+    Owned<IFileIO> io;
+    offset_t fileSize = 0;
+    size32_t blockSize = 0;
+    void *buffer = nullptr;
+    offset_t lastReadPos = (offset_t)-1;
+    MemoryBuffer mb;
+public:
+    CBlockedFileIO(IFileIO *_io, size32_t _blockSize) : io(_io), blockSize(_blockSize)
+    {
+        fileSize = io->size();
+        buffer = mb.reserveTruncate(blockSize);
+    }
+    virtual size32_t read(offset_t pos, size32_t len, void *data) override
+    {
+        if (len > blockSize || pos+len > fileSize || len==0)
```

Comment:
Do you need the fileSize checking?  It means that you may perform an extra stat call which will add latency if it is on blob storage.  It should be possible to change the code so it copes with reading at the end of the file.
Also len==0 test is not needed.

Reply:
agreed, refactored to avoid need for fileSize